update workflow deployment (#966)
microsoft/AzureTRE: docs/tre-admins/setup-instructions/workflows.md

# GitHub Actions workflows (CI/CD)

To deploy the Azure TRE using GitHub workflows, create a fork of the repository.

> Deployment is done using the `/.github\workflows\deploy_tre.yml` workflow. This method is also used to deploy the dev/test environment for the original AzureTRE repository.

Reviewer: ```suggestion
Deployment is done using the `/.github\workflows\deploy_tre.yml` workflow. This method is also used to deploy the dev/test environment for the original Azure TRE repository.
```

## Setup instructions

Before you can run the `deploy_tre.yml` pipeline there are some one-time configuration steps that we need to do, similar to the Pre-deployment steps for manual deployment.

!!! tip
    In some of the steps below, you are asked to configure repository secrets. Follow the [GitHub guide](https://docs.github.com/en/actions/security-guides/encrypted-secrets) on creating repository secrets if you are unfamiliar with this step.

> 1. Create a Service Principal for the subscription so that the workflow can provision Azure resources.

Reviewer: ```suggestion
1. Create a service principal for the subscription so that the workflow can provision Azure resources.
```

> 1. Decide on a TRE ID and Azure Resource Location

Reviewer: ```suggestion
1. Decide on a TRE ID and the location for the Azure resources
```

1. Create app registrations for API authentication
1. Create app registrations and a user for the E2E tests
1. Create a workspace app registration for setting up workspaces (for the E2E tests)
1. Create a Teams WebHook for deployment notifications
1. Configure repository secrets
1. Deploy the TRE using the workflow

> ### Create a Service Principal for provisioning resources

Reviewer: ```suggestion
### Create a service principal for provisioning resources
```

1. Login to Azure

  Log in to Azure using `az login` and select the Azure subscription you wish to deploy Azure TRE to:

  ```cmd
  az login
  az account list
  az account set --subscription <subscription ID>
  ```

  !!! caution
      When running locally the credentials of the logged-in user will be used to deploy the infrastructure. Hence, it is essential that the user has enough permissions to deploy all resources.

  See [Sign in with Azure CLI](https://docs.microsoft.com/cli/azure/authenticate-azure-cli) for more details.

> 1. Create a Service Principal

Reviewer: ```suggestion
1. Create a service principal
```

  A service principal needs to be created to authorize CI/CD workflows to provision resources for the TRE workspaces and workspace services.

  Create a main service principal with "**Owner**" role:

  ```cmd
  az ad sp create-for-rbac --name "sp-aztre-core" --role Owner --scopes /subscriptions/<subscription_id> --sdk-auth
  ```

  !!! caution
      Save the JSON output locally - as you will need it later for the `AZURE_CREDENTIALS` secret

1. Configure the AZURE_CREDENTIALS repository secret

  | <div style="width: 230px">Secret name</div> | Description |
  | ----------- | ----------- |

>   | `AZURE_CREDENTIALS` | The json output from the previous step |

Reviewer: ```suggestion
  | `AZURE_CREDENTIALS` | The JSON output from the previous step |
```

> ### Decide on a TRE ID and Azure Resource Location

Reviewer: ```suggestion
### Decide on a TRE ID and Azure resources location
```

Configure the TRE ID and LOCATION repository secrets

| <div style="width: 230px">Secret name</div> | Description |
| ----------- | ----------- |
| `TRE_ID` | A globally unique identifier. `TRE_ID` can be found in the resource names of the Azure TRE instance; for example, a `TRE_ID` of `tre-dev-42` will result in a resource group name for Azure TRE instance of `rg-tre-dev-42`. This must be less than 12 characters. Allowed characters: Alphanumeric, underscores, and hyphens. |
| `LOCATION` | The Azure location (region) for all resources. E.g. `westeurope` |

### Create app registrations for API authentication

Follow the instructions to run the **app registration script** in the [Authentication and Authorization document](../auth.md#app-registrations). Use the values for TRE ID and LOCATION from above.

Configure the TRE API and Swagger UI repository secrets

| <div style="width: 230px">Secret name</div> | Description |
| ----------- | ----------- |
| `AAD_TENANT_ID` | The tenant ID of the Azure AD. |

> | `SWAGGER_UI_CLIENT_ID` | The application (client) ID of the TRE Swagger UI Service Principal. |

Reviewer: ```suggestion
| `SWAGGER_UI_CLIENT_ID` | The application (client) ID of the TRE Swagger UI app registration. |
```

> | `API_CLIENT_ID` | The application (client) ID of the TRE API Service Principal. |

Reviewer: ```suggestion
| `API_CLIENT_ID` | The application (client) ID of the TRE API app. |
```

> | `API_CLIENT_SECRET` | The application password (client secret) of the TRE API Service Principal. |

Reviewer: ```suggestion
| `API_CLIENT_SECRET` | The application password (client secret) of the TRE API app. |
```

### Create an app registration and a user for the E2E tests

Follow the instructions to [create an app registration and a test user for the E2E tests in the Authentication and Authorization](../auth.md#tre-e2e-test) document.

Configure the E2E Test repository secrets

| <div style="width: 230px">Secret name</div> | Description |
| ----------- | ----------- |

> | `TEST_APP_ID` | The application (client) ID of the E2E Test app Service Principal |

Reviewer: ```suggestion
| `TEST_APP_ID` | The application (client) ID of the E2E Test app |
```

| `TEST_USER_NAME` | The username of the E2E Test User |
| `TEST_USER_PASSWORD` | The password of the E2E Test User |

### Create a workspace app registration for setting up workspaces (for the E2E tests)

Follow the [instructions to create a workspace app registration](../auth.md#workspaces) (used for the E2E tests) - and make the E2E test user a **WorkspaceOwner** for the app registration.

Configure the TEST_WORKSPACE_APP_ID repository secret

| <div style="width: 230px">Secret name</div> | Description |
| ----------- | ----------- |

> | `TEST_WORKSPACE_APP_ID` | The application (client) ID of the Workspaces app Service Principal. |

Reviewer: ```suggestion
| `TEST_WORKSPACE_APP_ID` | The application (client) ID of the Workspaces app. |
```
